src/snowflake/snowpark/_internal/packaging_utils.py

```diff
@@ -0,0 +1,376 @@
+                        if (
+                            os.path.exists(record_entry_full_path)
+                            and directory in record_entry_full_path
+                            and len(record_entry) > 0
+                        ):
+                            included_record_entries.append(record_entry)
+
+                    # Create Requirement objects and store in map
+                    package_name_to_record_entries_map[
+                        Requirement.parse(package)
+                    ] = included_record_entries
+
+    return package_name_to_record_entries_map
+
+
+def identify_supported_packages(
+    packages: List[Requirement],
+    valid_packages: Dict[str, List[str]],
+    native_packages: Set[str],
+) -> Tuple[List[Requirement], List[Requirement], List[Requirement]]:
+    """
+    Detects which `packages` are present in the Snowpark Anaconda channel using the `valid_packages` mapping.
+    If a package is a native dependency (belongs to `native_packages` set) and supported in Anaconda, we switch to
+    the latest available version in Anaconda.
+
+    Note that we also update the `native_packages` set to reflect genuinely problematic native dependencies, i.e.
+    packages that are not present in Anaconda and are likely to cause errors.
+
+    Args:
+        packages (List[Requirement]): List of python packages.
+        valid_packages (Dict[str, List[str]): Mapping from package name to a list of versions available on the Anaconda
+        channel.
+        native_packages (Set[str]): Set of native dependency package names.
+
+    Returns:
+        Tuple[List[Requirement], List[Requirement], List[Requirement]]: Tuple containing dependencies that are present
+        in Anaconda, dependencies that should be dropped from the package list and dependencies that should be added.
+    """
+
+    supported_dependencies: List[Requirement] = []
+    dropped_dependencies: List[Requirement] = []
+    new_dependencies: List[Requirement] = []
+
+    for package in packages:
+        package_name: str = package.name
+        package_version_required: Optional[str] = (
+            package.specs[0][1] if package.specs else None
+        )
+
+        if package_name in valid_packages:
+            # Detect supported packages
+            if (
+                package_version_required is None
+                or package_version_required in valid_packages[package_name]
+            ):
+                supported_dependencies.append(package)
+
+            # Native packages should be anaconda dependencies, even if the requested version is not available.
+            elif package_name in native_packages:
+                _logger.warning(
+                    f"Package {package_name}(version {package_version_required}) is an unavailable native "
+                    f"dependency, switching to latest available version "
```

Comment:
```suggestion
                    f"Package {package_name}(version {package_version_required}) has an unavailable native "
                    f"dependency, switching to latest available version "
```
Also, `package_version_required` can be `None` here, right?

Reply:
Fixed the version text;

"is" is more accurate because the package `package_name` is the native dependency (hard to know which requested package has this dependency because pip doesn't specify the association between dependencies and requested packages).